phoenix-core/src/main/java/org/apache/phoenix/compile/QueryCompiler.java

```diff
@@ -147,21 +169,146 @@ public QueryCompiler(PhoenixStatement statement, SelectStatement select, ColumnR
+            ProjectedColumnExpression expression = new ProjectedColumnExpression(tbl.getColumns().get(i), tbl.getColumns(), i, tbl.getColumns().get(i).getExpressionStr());
+            projectedColumns.add(new ExpressionProjector(tbl.getColumns().get(i).getName().getString(), tbl.getName().getString(), expression, true));
+        }
+        RowProjector rowProjector = new RowProjector(projectedColumns, 100, true);
+        TupleProjector tupleProjector = new TupleProjector(rowProjector);
+        plan = new TupleProjectionPlan(plan, tupleProjector, null);
+        return plan;
+    }
+    private boolean containOrderBy(SelectStatement select) {
+        if (select.getOrderBy() != null && !select.getOrderBy().isEmpty()) 
+            return true;
+        else
+            return false;
+    }
+
+    private QueryPlan compileUnionAll(SelectStatement select) throws SQLException { 
+        List<SelectStatement> unionAllSelects = select.getSelects();
+        List<QueryPlan> plans = new ArrayList<QueryPlan>();
         StatementContext context = new StatementContext(statement, resolver, scan, sequenceManager);
-        if (select.isJoin()) {
-            select = JoinCompiler.optimize(statement, select, resolver);
-            if (this.select != select) {
-                ColumnResolver resolver = FromCompiler.getResolverForQuery(select, statement.getConnection());
-                context = new StatementContext(statement, resolver, scan, sequenceManager);
+
+        checkForOrderByLimitInUnionAllSelect(select);
+        QueryPlan plan = compileSelect(select);
+        plan = buildTupleProjectPlan(plan);
+        plans.add(plan);
+        OrderBy orderBy = OrderBy.EMPTY_ORDER_BY;
+        boolean containOrderBy = false;
+        int numSelects = unionAllSelects.size();
+        for (int i=0; i < numSelects; i++ ) {
+            if (i < numSelects-1)
+                checkForOrderByLimitInUnionAllSelect(unionAllSelects.get(i));
+            else if (i == numSelects-1) {
+                containOrderBy = containOrderBy(unionAllSelects.get(i));
             }
-            JoinTable joinTable = JoinCompiler.compile(statement, select, context.getResolver());
-            return compileJoinQuery(context, binds, joinTable, false, false, null);
-        } else {
-            return compileSingleQuery(context, select, binds, false, true);
+            plan = compileSelect(unionAllSelects.get(i));
+            if (containOrderBy) {
+                orderBy = plan.getOrderBy();
+                unionAllSelects.get(i).removeOrderBy();
+                plan = compileSelect(unionAllSelects.get(i));
+            } 
+            plan = buildTupleProjectPlan(plan);
+            plans.add(plan);
         }
+        checkProjectionNumAndTypes(plans);
+        plan =  new UnionPlan(context, plan.getStatement(), plan.getTableRef(), plan.getProjector(), plan.getLimit(), orderBy, parallelIteratorFactory, plan.getGroupBy(), plans);
```

Comment:
OrderBy is used by union all plan to get Order by information. Cannot be null if it has order by clause.

When building the RowProjector, which "select" you referred to here. There are many selects in union all. Not matter which one you choose, effectively you pick one of the selects to build the RowProjector. It is the problem I have mentioned. And that is why I pass down RowProjector to result sets.
RowProjector rowProjector = ProjectorCompiler.compile(context, select, groupBy, Collections.emptyList());

From: maryannxue <notifications@github.com<mailto:notifications@github.com>>
Reply-To: apache/phoenix <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Thursday, March 5, 2015 at 7:43 PM
To: apache/phoenix <phoenix@noreply.github.com<mailto:phoenix@noreply.github.com>>
Cc: Alicia Shu <ashu@hortonworks.com<mailto:ashu@hortonworks.com>>
Subject: Re: [phoenix] Phoenix-1580 union all impl (#39)


In phoenix-core/src/main/java/org/apache/phoenix/compile/QueryCompiler.java<https://github.com/apache/phoenix/pull/39#discussion_r25924080>:

>          }
> +        checkProjectionNumAndTypes(plans);
> +        plan =  new UnionPlan(context, plan.getStatement(), plan.getTableRef(), plan.getProjector(), plan.getLimit(), orderBy, parallelIteratorFactory, plan.getGroupBy(), plans);


Should not pass the inner plan's attributes to UnionPlan, instead you should create its own tableRef, rowProjector, limit, orderBy, groupBy, etc.

For orderBy, groupBy and limit, I think we should just leave them as NULL. And the tableRef and rowProjector is what really matters here.

Since we have already used TupleProjectionPlan earlier, here the tableRef and rowProjector is more like a flat dummy one.

We should create a table like (pretty much what createTempTableForUnionAllResultResolver() does):
List columns = new ArrayList();
// we'll use one of inner plan's row projector just for type information.
for (ColumnProjector colProjector : plan.getProjector().getColumnProjectors()) {
PColumn column = new PColumnImpl(colProject.getName(), ..., colProjector.getExpression().getType(), ...);
}
PTable tempTable = PTableImpl.makePTable("dummyName", "dummySchemaName", PTableType.SUBQUERY, columns);
TableRef tableRef = new TableRef(null, tempTable, 0);

GroupBy groupBy = GroupBy.EMPTY_GROUP_BY;
OrderBy orderBy = OrderBy.EMPTY_ORDER_BY;
context.setResolver(FromCompiler.getResolver(tableRef));
RowProjector rowProjector = ProjectorCompiler.compile(context, select, groupBy, Collections.emptyList());

plan = new UnionPlan(context, this.statement, tableRef, rowProjector, null, orderBy, parallelIteratorFactory, groupBy, plans);

-
Reply to this email directly or view it on GitHub<https://github.com/apache/phoenix/pull/39/files#r25924080>.